### PR TITLE
Add v2 permissions cache implementation

### PR DESCRIPTION
**Pull request by CarlHembrough**
- Poll in the background for permissions updates and then keep an in memory cache of the permissions model for a set period of time.
- Once the max cache validity time is reached we drop the model, thereby removing all permissions which will result in all permissions decisions returning false.

CachingStore wraps an existing permissions store implementation to provide an in memory cache layer.

### CODE REVIEW

File: v2/permissions/cache.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	go func() {
+		defer close(c.expiryCheckerClosed)
+		ticker := time.NewTicker(checkInterval)
+
+		for {
+			select {
+			case <-ticker.C:
+				c.CheckCacheExpiry(ctx, maxCacheTime)
+			case <-c.closing:
+				ticker.Stop()
```
Comment (daiLlew) [inline v2/permissions/cache.go:71]: I think you might also need to close `ticker.C`. https://pkg.go.dev/time#Ticker.Stop

File: v2/permissions/cache.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	return &CachingStore{
+		underlyingStore:     underlyingStore,
+		closing:             make(chan struct{}),
+		expiryCheckerClosed: make(chan struct{}),
+		cacheUpdaterClosed:  make(chan struct{}),
+	}
+}
+
+// GetPermissionsBundle returns the cached permission data, or an error if it's not cached.
+func (c *CachingStore) GetPermissionsBundle(ctx context.Context) (*Bundle, error) {
```
Comment (daiLlew) [inline v2/permissions/cache.go:35]: Does this need to return an error if the bundle is not cached? Could it return nil instead? I assume that if its not cached the next step is to retrieve the data and cache it? If so err feels wrong as its technically not an error - its a valid state.

File: v2/permissions/cache.go (hunk 1 above)
Comment (CarlHembrough) [reply v2/permissions/cache.go:71]: I think this is by design that the ticker channel is not closed https://github.com/golang/go/issues/2650#issuecomment-66061180

Having the ticker stop and function return on `c.closing` should allow it to be fully garbage collected? I couldn't see any examples where it suggested the channel should be closed.

File: v2/permissions/cache.go (hunk 2 above)
Comment (CarlHembrough) [reply v2/permissions/cache.go:35]: Yes this is by design. The cache should always be populated. If it's not, it means that the (background) cache update has failed, and the existing cache data has expired.

File: v2/permissions/cache.go (hunk 1 above)
Comment (daiLlew) [reply v2/permissions/cache.go:71]: Sorry I misread the godoc - They are saying that stop stops the ticker but if you care about go routines still accessing it while its stopping then you need to close the channel - at least I think thats what they mean.

For this that doesn't appear to be a problem so resolving comment.

**Review by daiLlew — APPROVED** [review]

**Pull request merged**